File: frontend/src/features/masters/models/invoice.ts
import type { PaginatedResponse } from './common';

export type InvoiceStatus = 'Open' | 'Payment Cleared' | 'Settled';

export interface InvoiceLine {
  id: string;
  product_detail_id: string;
  product_child_code: string | null;
  product_name: string | null;
  quantity: number | null;
  line_amount: number | null;
  vat_gst_amount: number | null;
}

export interface Invoice {
  id: string;
  invoice_number: string;
  invoice_date: string;
  vendor_id: string;
  vendor_name: string;
  customer_id: string | null;
  customer_name: string | null;
  bill_amount_excl_gst: number;
  bill_amount_incl_tax: number | null;
  amount_deducted: number;
  tds_value: number;
  due_date: string | null;
  payment_clearing_date: string | null;
  sap_clearing_doc_no: string | null;
  /** invoice_status is the canonical field returned by the backend */
  invoice_status: InvoiceStatus;
  /** status mirrors invoice_status for frontend convenience */
  status: InvoiceStatus;
  lines: InvoiceLine[];
  created_at: string;
  modified_at: string;
  created_by: string | null;
  modified_by: string | null;
}

export interface InvoiceListResponse extends PaginatedResponse<Invoice> {}

export interface CreateInvoiceLineRequest {
  product_detail_id: string;
  quantity?: number;
  line_amount?: number;
  vat_gst_amount?: number;
}

export interface CreateInvoiceRequest {
  invoice_number: string;
  invoice_date: string;
  vendor_id: string;
  customer_id: string;
  bill_amount_excl_gst: number;
  bill_amount_incl_tax?: number;
  amount_deducted?: number;
  tds_value?: number;
  lines: CreateInvoiceLineRequest[];
}

export interface PaymentUpdateRequest {
  payment_clearing_date: string;
  sap_clearing_doc_no?: string;
}

export interface EligibilityCheck {
  agreement_valid: boolean;
  agreement_message: string;
  mapping_valid: boolean;
  mapping_message: string;
  status_valid: boolean;
  status_message: string;
  overall_eligible: boolean;
}

File: backend/src/domain/entities/masters/invoice.py
"""
Invoice Header and Invoice Line domain entities.

An Invoice is a two-level aggregate: an Invoice Header owns one or more Invoice
Lines. The header captures billing totals, the vendor/customer relationship, and
payment-clearing lifecycle fields, while each line references a billable Product
Detail.
"""

from dataclasses import dataclass, field
from datetime import date
from decimal import Decimal
from uuid import UUID

from src.domain.entities.base_entity import BaseEntity
from src.domain.enums.masters import InvoiceStatus


@dataclass
class InvoiceLineEntity(BaseEntity):
    """
    Invoice Line entity (a single billed product row on an invoice).

    Business Rules:
    - Must belong to an existing Invoice Header (``invoice_header_id``).
    - Must reference an existing Product Detail (``product_detail_id``).
    - Quantity and amounts, when supplied, are non-negative (validated at the
      service level).
    """

    invoice_header_id: UUID | None = field(default=None)
    product_detail_id: UUID | None = field(default=None)
    quantity: Decimal | None = field(default=None)
    line_amount: Decimal | None = field(default=None)
    vat_gst_amount: Decimal | None = field(default=None)


@dataclass
class InvoiceHeaderEntity(BaseEntity):
    """
    Invoice Header aggregate root.

    Business Rules:
    - Invoice Number must be unique across invoice headers (enforced at the
      repository/DB level).
    - Must reference an existing Vendor and Customer.
    - ``bill_amount_excl_gst`` is required; ``amount_deducted`` and ``tds_value``
      default to 0.
    - New invoices default to ``Open`` status.
    - Deleting a header cascades to its lines (DB ``ON DELETE CASCADE``).
    """

    invoice_number: str = field(default="")
    invoice_date: date | None = field(default=None)
    vendor_id: UUID | None = field(default=None)
    customer_id: UUID | None = field(default=None)
    bill_amount_excl_gst: Decimal = field(default=Decimal("0"))
    bill_amount_incl_tax: Decimal | None = field(default=None)
    amount_deducted: Decimal = field(default=Decimal("0"))
    tds_value: Decimal = field(default=Decimal("0"))
    due_date: date | None = field(default=None)
    payment_clearing_date: date | None = field(default=None)
    sap_clearing_document_no: str | None = field(default=None)
    invoice_status: InvoiceStatus = field(default=InvoiceStatus.Open)
    lines: list[InvoiceLineEntity] = field(default_factory=list)
